accepted/0000-workspaces-link-phase-3.md

```diff
@@ -0,0 +1,164 @@
+- Yarn Issue: https://github.com/yarnpkg/yarn/issues/3294
+
+# Yarn workspaces phase 3: linking workspaces to each other
+
+## Summary
+
+A continuation of https://github.com/yarnpkg/rfcs/pull/60.
+Ability for workspaces to refer each other when testing packages in integration.
+
+## Motivation
+
+People tend to split larger projects into self contained packages that are published to npm independently. The workspaces feature is being developed for Yarn to address this workflow.
+
+In particular, testing packages that refer other packages from the same codebase can be difficult because Node.js and front end bundling tools would look up the referred packages in node_modules folder as it should be installed from npm registry.
+
+Yarn Workspaces need to enable referring other local packages the same way when local packages are in development mode (source of truth is the package source code) and in production mode (source of truth is the package installed from npm).
+
+## Detailed design
+
+The structure of the source code is following
+
+```
+| jest/
+| ---- package.json
+| ---- packages/
+| -------- jest-matcher-utils/
+| ------------ packjage.json
+| -------- jest-diff/
+| ------------ packjage.json
+...
+```
+
+Top level package.json is like
+
+```
+{
+  "private": true,
+  "name": "jest",
+  "devDependencies": {
+  },
+  "workspaces": [
+    "packages/*"
+  ]
+}
+```
+jest-matcher-utils (workspace referred by another one)
+```
+{
+  "name": "jest-matcher-utils",
+  "description": "A set of utility functions for jest-matchers and related packages",
+  "version": "20.0.3",
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/facebook/jest.git"
+  },
+  "license": "BSD-3-Clause",
+  "main": "build/index.js",
+  "browser": "build-es5/index.js",
+  "dependencies": {
+    "chalk": "^1.1.3",
+    "pretty-format": "^20.0.3"
+  }
+}
+```
+
+jest-diff (workspace that refers jest-matcher-utils)
+```
+{
+  "name": "jest-diff",
+  "version": "20.0.3",
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/facebook/jest.git"
+  },
+  "license": "BSD-3-Clause",
+  "main": "build/index.js",
+  "browser": "build-es5/index.js",
+  "dependencies": {
+    "chalk": "^1.1.3",
+    "diff": "^3.2.0",
+    "**jest-matcher-utils**": "^20.0.3",
+    "pretty-format": "^20.0.3"
+  }
+}
+```
+
+When user runs yarn install, this folder structure of the Workspace gets created
+```
+| jest/
+| ---- node_modules/
+| -------- chalk/
+| -------- diff/
+| -------- pretty-format/
+| ---- package.json
+| ---- packages/
+| -------- jest-matcher-utils/
+| ------------ node_modules/ (empty, all dependencies hoisted to the root)
+| ------------ package.json
+| -------- jest-diff/
+| ------------ node_modules/
+| ---------------- **jest-matcher-utils**/  (symlink) -> ../jest-matcher-utils
+| ------------ package.json
+...
+```
+
+### Dependencies and version matching
+
+Yarn would only link workspaces to each other if they match semver conditions.
```

Comment:
I don't think it should always warn. This breaks Babel's use case where it is self-hoisting and wants to use an older version intentionally.